docs/basics/101-146-providers.rst

```diff
@@ -0,0 +1,86 @@
+.. _providers:
+
+Configuring new data providers
+------------------------------
+
+DataLad can download files via the ``http``, ``https``, ``ftp``, and ``s3``
+protocol. If data retrieval requires authentication, DataLad's provides an interface to
+query, request, and store the most common type of credentials. An example of this
+is shown in the usecase :ref:`usecase_HCP_dataset`: :command:`datalad get` will
+ask for AWS S3 credential to retrieve data that is stored in S3 buckets from
+the command line once, and -- given valid credentials -- download the files
+and all future files from this bucket right away. There
+are a number of natively supported types of authentication (a list is given in
+the hidden section below) and out-of-the box access to a broad range
+of providers such as `S3 <https://aws.amazon.com/s3/?nc1=h_ls>`_ or
+`LORIS <https://loris.ca/>`_.
+If you attempt to retrieve data from supported providers that require authentication,
+you will be prompted for your credentials from the command line at the time of
+the first download, and subsequent downloads handle authentication in the background.
+
+If a provider requires authentication but it is not known to DataLad yet, the
+download will fail. However, the provider can be configured by
+the user to enable interactions with it.
+With a provider configuration in place, commands such as :command:`datalad download-url`
+or :command:`datalad add-urls` can work with urls of custom providers, and
+:command:`datalad get` is enabled to retrieve file contents from these sources.
+The configuration can either be done in the terminal upon a prompt from the
```

Comment:
do you mean s/configuration/credential entry/ ? Is there a way to configure "providers" on the fly?

Reply:
Yes (I only know from trying it blindly). I creates the same file in the same location as you have. Here's how looks in the terminal when downloading without the cfg file in place:

```
datalad download-url  https://nmrxnat.ime.kfa-juelich.de:8443/xnat/data/projects/1000BRAINS/subjects/XNATINM4_S03059/experiments/XNATINM4_E04029/scans/1/resources/85041/files/0015911870_1.3.12.2.1107.5.2.32.35135.2011102112040130362336780.dcm
[INFO   ] Downloading 'https://nmrxnat.ime.kfa-juelich.de:8443/xnat/data/projects/1000BRAINS/subjects/XNATINM4_S03059/experiments/XNATINM4_E04029/scans/1/resources/85041/files/0015911870_1.3.12.2.1107.5.2.32.35135.2011102112040130362336780.dcm' into '/tmp/xnat2/' 
Authenticated access to https://nmrxnat.ime.kfa-juelich.de:8443/xnat/data/projects/1000BRAINS/subjects/XNATINM4_S03059/experiments/XNATINM4_E04029/scans/1/resources/85041/files/0015911870_1.3.12.2.1107.5.2.32.35135.2011102112040130362336780.dcm has failed.
Would you like to setup a new provider configuration to access url? (choices: [yes], no): yes

New provider name
Unique name to identify 'provider' for https://nmrxnat.ime.kfa-juelich.de:8443/xnat/data/projects/1000BRAINS/subjects/XNATINM4_S03059/experiments/XNATINM4_E04029/scans/1/resources/85041/files/0015911870_1.3.12.2.1107.5.2.32.35135.2011102112040130362336780.dcm [nmrxnat.ime.kfa-juelich.de]: xnat

New provider regular expression
A (Python) regular expression to specify for which URLs this provider should be used [https://nmrxnat\.ime\.kfa\-juelich\.de:8443/xnat/data/projects/1000BRAINS/subjects/XNATINM4_S03059/experiments/XNATINM4_E04029/scans/1/resources/85041/files/0015911870_1\.3\.12\.2\.1107\.5\.2\.32\.35135\.2011102112040130362336780\.dcm]: https://nmrxnat.ime.kfa-juelich.de:8443/xnat/.*

Authentication type
What authentication type to use (choices: aws-s3, bearer_token, html_form, http_auth, http_basic_auth, http_digest_auth, loris-token, nda-s3, none, xnat): http_basic_auth

Credential
What type of credential should be used? (choices: aws-s3, loris-token, nda-s3, token, [user_password]): 

Save provider configuration file
Following configuration will be written to /home/adina/.config/datalad/providers/xnat.cfg:
# Provider configuration file created to initially access
# https://nmrxnat.ime.kfa-juelich.de:8443/xnat/data/projects/1000BRAINS/subjects/XNATINM4_S03059/experiments/XNATINM4_E04029/scans/1/resources/85041/files/0015911870_1.3.12.2.1107.5.2.32.35135.2011102112040130362336780.dcm

[provider:xnat]
url_re = https://nmrxnat.ime.kfa-juelich.de:8443/xnat/.*
authentication_type = http_basic_auth
# Note that you might need to specify additional fields specific to the
# authenticator.  Fow now "look into the docs/source" of <class 'datalad.downloaders.http.HTTPBasicAuthAuthenticator'>
# http_basic_auth_
credential = xnat

[credential:xnat]
# If known, specify URL or email to how/where to request credentials
# url = ???
type = user_password
 (choices: [yes], no): yes

You need to authenticate with 'xnat' credentials.
user: a.wagner

password: 
password (repeat): 
[INFO   ] http session: Authenticating into session for https://nmrxnat.ime.kfa-juelich.de:8443/xnat/data/projects/1000BRAINS/subjects/XNATINM4_S03059/experiments/XNATINM4_E04029/scans/1/resources/85041/files/0015911870_1.3.12.2.1107.5.2.32.35135.2011102112040130362336780.dcm 
https://nmrxnat.ime.kfa-juelich.de:8443/xnat/data/projects/1000BRAINS/subjects/XNATINM4_S03059/experiments/XNATINM4_E04029/scans/1/resources/85041/files/0015911870_1.3.12.2.1107.5.2.32.35135.2011102112040130362336780.dcm:   0%| | 0.00/611k                                                                                                                                                                                                                                                 download_url(ok): /tmp/xnat2/0015911870_1.3.12.2.1107.5.2.32.35135.2011102112040130362336780.dcm (file)
add(ok): 0015911870_1.3.12.2.1107.5.2.32.35135.2011102112040130362336780.dcm (file)
save(ok): . (dataset)
action summary:
  add (ok: 1)
  download_url (ok: 1)
  save (ok: 1)
```